### PR TITLE
Per week tests

### PR DESCRIPTION
**Pull request by leepatrickworrall**
Tests and new feature implemented into the code.

### CODE REVIEW

File: ketchup/main.py [hunk 1]
```diff
@@ -75,46 +77,50 @@ def run():
...
+    weeks_required = math.ceil(episodes / total_episodes_per_week)
+
     # Time to math!
-    watch_schedule = wsb.calculate_watch_schedule(
-        episodes, end_date, start_date, per_day_limit
+    watch_schedule = dc.calculate_watch_schedule(
+        episodes, end_date, start_date, per_day_limit, total_episodes_per_week
     )
-    print(f"Wow, that's only {watch_schedule['days']} days between today and then !!!")
+    print(f"Wow, that's only {watch_schedule['days']} days between today and then!")
```
Comment (chellygel) [inline ketchup/main.py:88]: how dare you remove my excitement and shock (lol jk, its fine)

File: ketchup/main.py [hunk 2]
```diff
@@ -75,46 +77,50 @@ def run():
...
-            "to complete the show on time!"
+            f"To complete the show on time, dedicating {days_per_week_watching} days per week and watching "
+            f"{per_day_limit} episodes per day, it will take approximately {weeks_required} week(s)."
         )
         if watch_schedule["extra_episodes"]:
             print(
-                f"You will have {watch_schedule['extra_episodes']} extra,"
-                "you can add an additional watch day or tack them on to one of your other days."
+                f"You will have {watch_schedule['extra_episodes']} extra, "
+                f"you can add an additional watch day or tack them on to one of your other days."
```
Comment (chellygel) [inline ketchup/main.py:98]: there's nothing to format here so no need for f string

File: ketchup/main.py [hunk 3]
```diff
@@ -75,46 +77,50 @@ def run():
...
             )
     else:
-        print("I'm sorry, but that limit and end date make it impossible!")
-        print("Don't worry, consider these options instead:")
+        print(f"I'm sorry, but that limit and end date make it impossible!")
```
Comment (chellygel) [inline ketchup/main.py:101]: same with these f's we dont need them

File: ketchup/main.py [hunk 4]
```diff
@@ -1,9 +1,9 @@
 import datetime
-
+import math
 import dateparser
 import datetime as dt
 from ketchup.luffy import LUFFY
-import ketchup.watch_schedule_builder as wsb
+import ketchup.watch_schedule_builder as dc
```
Comment (chellygel) [inline ketchup/main.py:6]: i think you want to change this back to wsb

File: ketchup/main.py [hunk 5]
```diff
@@ -37,35 +37,37 @@ def run():
...
+        start_date_prompt = "What day will you start? (Ex: 15 Aug 2023):\n"
         start_date = get_date(start_date_prompt)
 
-    per_day_limit = (
-        input("What is the limit of episodes you will watch per day? [0+] ") or 0
+    per_day_limit = int(
+        input("What is the limit of episodes you will watch per day? [0+]\n") or 0
     )
 
+    days_per_week_watching = int(input("How many days a week do you plan to watch?:\n"))
```
Comment (chellygel) [inline ketchup/main.py:69]: perhaps we need an explainer here?

"Some people may not be able to watch 7 days a week, how many days per week will you be available to watch episodes?" or something similar? thoughts?

File: ketchup/main.py [hunk 6]
```diff
@@ -75,46 +77,50 @@ def run():
...
 
+    total_episodes_per_week = per_day_limit * days_per_week_watching
+
+    weeks_required = math.ceil(episodes / total_episodes_per_week)
+
     # Time to math!
-    watch_schedule = wsb.calculate_watch_schedule(
-        episodes, end_date, start_date, per_day_limit
+    watch_schedule = dc.calculate_watch_schedule(
+        episodes, end_date, start_date, per_day_limit, total_episodes_per_week
```
Comment (chellygel) [inline ketchup/main.py:86]: Ah this makes sense, nice.

File: ketchup/watch_schedule_builder.py [hunk 7]
```diff
@@ -45,4 +48,11 @@ def calculate_watch_schedule(episodes, end_date, start_date, per_day_limit=None)
             # It isn't possible to watch with that limit so...
             response["is_possible"] = False
 
+    if per_week_limit:
```
Comment (chellygel) [inline ketchup/watch_schedule_builder.py:51]: So lets write an additional test (in the separate test file) that would effect this log.

Consider something like testing the math itself... a test case could be like:

Total episodes -> 15
Episodes Per Day limit -> 5 eps
Days per week limit -> 3 days per week
total eps per week -> 15 episodes per week
start date -> August 21, 2023
end date -> September 30, 2023

assert that math is correct  (only 1 week is required to watch the show)

Then do a different test where one thing changes:

**Total episodes -> 150**
Episodes Per Day limit -> 5 eps
Days per week limit -> 3 days per week
total eps per week -> 15 episodes per week
start date -> August 21, 2023
end date -> September 30, 2023

we know this case isnt possible in that time frame, so we should assert the response has changed.
You can assert on multiple things (this is a contentious topic in the industry, you can see some of that [here in this article](https://methodpoet.com/are-multiple-asserts-bad/). )   So we can assert :

```
response = {
        "base_eps_per_day": 0,
        "days": 0,
        "actual_watch_days": 0,
        "is_possible": True,
        "realistic_end_date": None,
        "extra_episodes": None,
    }
```

Each of these values are set as we expect. This way, the user who calls this function can get the response they'd expect. So you could assert that response['base_eps_per_day'] is equal to x, same with days, etc.

Now you can start to consider multiple cases for this function ! 

Ideally, we want every line of this function tested, and the best way to do that is to activate those if statements via different input variables. Does this make sense?

File: test/test_watch_schedule_builder/test_per_week_limit.py [hunk 8]
```diff
@@ -0,0 +1,29 @@
+from datetime import date
+from ketchup import watch_schedule_builder as wsb
+
+
+def test_per_week_limit_none():
```
Comment (chellygel) [inline test/test_watch_schedule_builder/test_per_week_limit.py:5]: i havent run these but they look roughly correct on glance. though this test is only testing that the "is_possible" response is set to true, not that anything else occurred. So we may need a bit more depth on this test specifically?

**Review by chellygel — CHANGES_REQUESTED** [review]
added some comments and reviewed tests specifically with some additional information.

File: ketchup/main.py (hunk 4 above)
Comment (leepatrickworrall) [reply ketchup/main.py:6]: Yeah, merge errors came through. Will resolve.

File: ketchup/main.py (hunk 5 above)
Comment (leepatrickworrall) [reply ketchup/main.py:69]: Trying to find a way to offer that without it being too verbose.

Maybe a user input to be taken before this one. Asking the user if they can/cannot dedicate time to watching an episode a day, and if they can't dedicate an episode daily, the user input for days_per_week_watching would be prompted.

File: ketchup/main.py (hunk 6 above)
Comment (leepatrickworrall) [reply ketchup/main.py:86]: Yeah, more than the per_week_limit as beforehand. It was even confusing me at times when testing the code.

The wording was "How many episodes would you like to watch a week?" and I would put 10, but the calculation is dividing the overall episodes by seven, for days in the week. So the outcome I expected was incorrect.

File: test/test_watch_schedule_builder/test_per_week_limit.py (hunk 8 above)
Comment (leepatrickworrall) [reply test/test_watch_schedule_builder/test_per_week_limit.py:5]: Yeah definitely I see what you mean. They were just baseline tests for myself, but as you said in a previous, we want to be able to test each line of code in the file.

**Comment by leepatrickworrall** [comment]
Archiving due to the code currently being outdated since the pull request was made.

**Pull request closed without merging**